Add troubleshooting guide for OOMKilled EDOT Collector pods using pprof (#270)
elastic/opentelemetry: docs/reference/edot-collector/troubleshooting/collector-oomkilled.md

---
navigation_title: EDOT Collector is out of memory
description: Diagnose and resolve out-of-memory issues in the EDOT Collector using Go’s Performance Profiler.
applies_to:
  stack:
  serverless:
    observability:
products:
  - id: cloud-serverless
  - id: observability
  - id: edot-collector
---

> # EDOT Collector is out of memory

Reviewer: I was wondering if we could prefix the title with "Issue:" or similar, to make it more identifiable everywhere (search, nav, etc.) as a tshooting topic.

Author: Does that make sense if we plan to put it under "Troubleshooting" eventually? The topics that live there don't follow that structure. 

We could do something like "Troubleshooting: EDOT Collector is out of memory", but won't it be a bit redundant in the current structure?

Reviewer: I'd follow this example: https://www.elastic.co/docs/troubleshoot/monitoring/high-memory-pressure

- Title with action verb ("Understand X" or "Troubleshoot Z")
- Navigation title similarly short and to the point

Author: How about:

- "Troubleshoot an out-of-memory EDOT Collector" for the title
- "EDOT Collector out of memory" for the navigation

Reviewer: Nav title could be even shorter: "Collector out of memory". Otherwise LGTM!

> If your EDOT Collector is being terminated with an `OOMKilled` status, this usually indicates a memory leak or sustained memory pressure. You can use the Performance Profiler (`pprof`) tool to collect and analyze memory profiles, helping you identify the root cause.

Reviewer: ```suggestion
If your EDOT Collector terminates with an `OOMKilled` status, this usually indicates a memory leak or sustained memory pressure. You can use the Performance Profiler (`pprof`) extension to collect and analyze memory profiles, helping you identify the root cause of the issue.
```

The Collector has the concept of extensions for this kind of meta tasks -- better use the specific term as you do later in the doc.

> If your EDOT Collector is being terminated with an `OOMKilled` status, this usually indicates a memory leak or sustained memory pressure. You can use the Performance Profiler (`pprof`) tool to collect and analyze memory profiles, helping you identify the root cause.

Reviewer: I think it's fine to write "pods" lowercase here, since we're not talking of an API object, as per K8s style guide: https://kubernetes.io/docs/contribute/style/style-guide/#use-upper-camel-case-for-api-objects

> If your EDOT Collector is being terminated with an `OOMKilled` status, this usually indicates a memory leak or sustained memory pressure. You can use the Performance Profiler (`pprof`) tool to collect and analyze memory profiles, helping you identify the root cause.

Author: Yup, I agree with keeping pods lowercase like the rest of the doc. Here is the current version with the comments applied:

> If your EDOT Collector pods terminate with an `OOMKilled` status, this usually indicates sustained memory pressure or potentially a memory leak due to an introduced regression or a bug. You can use the Performance Profiler (`pprof`) extension to collect and analyze memory profiles, helping you identify the root cause of the issue.

> If your EDOT Collector is being terminated with an `OOMKilled` status, this usually indicates a memory leak or sustained memory pressure. You can use the Performance Profiler (`pprof`) tool to collect and analyze memory profiles, helping you identify the root cause.

Author: Makes sense! See comment above for the updated paragraph.

> If your EDOT Collector is being terminated with an `OOMKilled` status, this usually indicates a memory leak or sustained memory pressure. You can use the Performance Profiler (`pprof`) tool to collect and analyze memory profiles, helping you identify the root cause.

Reviewer 2: Thank's!

## Symptoms

> 

Author: Good point, thank you both!

- EDOT Collector pod restarts with an `OOMKilled` status in Kubernetes.
- Memory usage steadily increases before the crash.

> - The collector's logs don't show clear errors before termination.

Reviewer: ```suggestion
- The Collector's logs don't show clear errors before termination.
```

Always capitalized when singular

## Resolution

> Enable runtime profiling using the `pprof` extension and then gather memory heap profiles from the affected pod:

Reviewer: ```suggestion
Turn on runtime profiling using the `pprof` extension and then gather memory heap profiles from the affected pod:
```

We prefer "turn on" or "activate" to avoid ableist language

> :::::{stepper}

Reviewer: Yay!

  pipelines:
    metrics:
      receivers: [ ... ]
      processors: [ ... ]
      exporters: [ ... ]
```

Restart the Collector after applying these changes. When the Daemonset is deployed again, spot the pod that is getting restarted.
::::

::::{step}
**Access the affected pod and collect a heap dump**

When a pod starts exhibiting high memory usage or restarts due to OOM, run the following to enter a debug shell:

```console
kubectl debug -it <collector-pod-name> --image=ubuntu:latest
```

In the debug container:

```console
apt update
apt install -y curl
curl http://localhost:1777/debug/pprof/heap > heap.out
```
::::

::::{step}
**Copy the heap file from the pod**

From your local machine, copy the heap file using:

```bash
kubectl cp <collector-pod-name>:heap.out ./heap.out -c <debug-container-name>
```
::::{note}
Replace `<debug-container-name>` with the name assigned to the debug container. Without the `-c` flag, Kubernetes will show the list of available containers.
::::
::::

::::{step}
**Convert the heap profile for analysis**

You can now generate a visual representation, for example PNG:

```bash
go tool pprof -png heap.out > heap.png
```
::::

## Best practices

> 

Author: Ok! Makes sense to keep the structure consistent

- Collect multiple heap profiles over time (for example, every few minutes) to observe memory trends before the crash.

> - Consider automating heap profile collection at intervals to observe trends over time.

Author: Not sure if we should mention automating the heap collection if we don't provide a guide for this - TBD.
